platform/darwin/src/MGLMapSnapshotter.mm

```diff
@@ -0,0 +1,163 @@
+    std::unique_ptr<mbgl::Actor<mbgl::MapSnapshotter::Callback>> snapshotCallback;
+}
+
+- (instancetype)initWithOptions:(MGLMapSnapshotOptions*)options;
+{
+    self = [super init];
+    if (self) {
+        _loading = false;
+        
+        mbgl::DefaultFileSource *mbglFileSource = [MGLOfflineStorage sharedOfflineStorage].mbglFileSource;
+        mbglThreadPool = mbgl::sharedThreadPool();
+        
+        std::string styleURL = std::string([options.styleURL.absoluteString UTF8String]);
+        
+        // Size; taking into account the minimum texture size for OpenGL ES
+        mbgl::Size size = {
+            static_cast<uint32_t>(MAX(options.size.width, 64)),
+            static_cast<uint32_t>(MAX(options.size.height, 64))
+        };
+        
+        float pixelRatio = MAX(options.scale, 1);
+        
+        // Camera options
+        mbgl::CameraOptions cameraOptions;
+        if (CLLocationCoordinate2DIsValid(options.camera.centerCoordinate)) {
+            cameraOptions.center = MGLLatLngFromLocationCoordinate2D(options.camera.centerCoordinate);
+        }
+        cameraOptions.angle = MAX(0, options.camera.heading) * mbgl::util::DEG2RAD;
+        cameraOptions.zoom = MAX(0, options.zoom);
+        cameraOptions.pitch = MAX(0, options.camera.pitch);
+        
+        // Region
+        mbgl::optional<mbgl::LatLngBounds> region;
+        if (!MGLCoordinateBoundsIsEmpty(options.region)) {
+            region = MGLLatLngBoundsFromCoordinateBounds(options.region);
+        }
+        
+        // Create the snapshotter
+        mbglMapSnapshotter = std::make_unique<mbgl::MapSnapshotter>(*mbglFileSource, *mbglThreadPool, styleURL, size, pixelRatio, cameraOptions, region);
+    }
+    return self;
+}
+
+- (void)startWithCompletionHandler:(MGLMapSnapshotCompletionHandler)completion;
+{
+    [self startWithQueue:dispatch_get_main_queue() completionHandler:completion];
+}
+
+- (void)startWithQueue:(dispatch_queue_t)queue completionHandler:(MGLMapSnapshotCompletionHandler)completion;
+{
+    if ([self isLoading]) {
+        NSDictionary *userInfo = @{NSLocalizedDescriptionKey: @"Already started this snapshotter"};
+        NSError *error = [NSError errorWithDomain:MGLErrorDomain code:1 userInfo:userInfo];
+        dispatch_async(queue, ^{
+           completion(nil, error); 
+        });
+        return;
+    }
+    
+    _loading = true;
+    
+    dispatch_async(queue, ^{
+        snapshotCallback = std::make_unique<mbgl::Actor<mbgl::MapSnapshotter::Callback>>(*mbgl::Scheduler::GetCurrent(), [=](std::exception_ptr mbglError, mbgl::PremultipliedImage image) {
+            _loading = false;
+            if (mbglError) {
+                NSString *description = @(mbgl::util::toString(mbglError).c_str());
+                NSDictionary *userInfo = @{NSLocalizedDescriptionKey: description};
+                NSError *error = [NSError errorWithDomain:MGLErrorDomain code:1 userInfo:userInfo];
```

Comment:
This arbitrary code should be replaced with a new value of MGLErrorCode defined in MGLTypes.h.